### ISSUE TITLE
Adding metadata to actions for middleware

### ISSUE THREAD

**Issue by xaviergonz** [mobxjs/mobx-state-tree#290]
I want to create a middleware that makes some actions be ignored on the client and be sent to the server (so they are are applied there and later replicated on the clients), while others should run on the clients directly.

Is there a way to mark some actions with some metadata (e.g. a decorator or something) so the middle ware can know how to act on each one instead of having to add lots of if statements on the object type + action name?

If not, is there at least a way to add a "metadata" section to models to add a list with action names or something so the middle ware would be able to read the section and act accordingly?

Also, is there a way to get the model out of an instance? something like call.object instanceof Todo or something equivalent

**Comment by mattiamanzati** [comment]
You can use getType(instance) to get the model out of an instance :)
About the decorators/metadata I think that #282 will address that because you could attach tags directly to the action itself, even using decorators! :)

**Comment by xaviergonz** [comment]
Awesome :D really looking forward to #282 then. I think I'll just use some array with action names set in the volatile state or some custom variable attached to the type for now then.

**Comment by mweststrate** [comment]
I think we still need something for this. If there is middleware for example for `undo` /`redo`, or `atomic` updates etc, you want to signal declaratively that you are interested in using such middleware for a certain action. I definitely see value in a proposal like this

**Comment by xaviergonz** [comment]
What about something like this:
```js
function redoDecorator(innerFunc) {
  return actionWithMetadata(innerFunc, { undoRedo: true}); // action with metadata comes from mobx and basically returns innerFunc + innerFunc.metadata = the object passed as parameter
}

.actions(self => ({
  someAction: redoDecorator(() => {
    // whatever the action has to do
  })
})
```

and ISerializedAction has a metadata field which is basically a clone / reference to the function.metadata if available

**Comment by mweststrate** [comment]
sorry, forgot to update this issue, but something like this is currently in
progress :)

Op ma 28 aug. 2017 om 21:33 schreef xaviergonz <notifications@github.com>:

**Comment by mweststrate** [comment]
will write down the proposed spec shortly, but here is a preview:

.actions(self => {
const takeA____ = process(function*() {
self.toilet.donate()
self.wipe()
self.wipe()
yield self.toilet.flush()
self.wipe()
self.wipe()
})

return {
// takeA____
takeA____: decorate(atomicAsync2, takeA____)
}
})

Op ma 28 aug. 2017 om 21:52 schreef Michel Weststrate <mweststrate@gmail.com

**Comment by xaviergonz** [comment]
Sweet! Any idea of how would the decorator itself look then?

**Comment by mweststrate** [comment]
Something like:

```
function undoRedo(call, next) {
   call.useUndoRedo = true // store meta info on the call
   return next(call)
}

function undoRedoManager(call, next) {
   if (call.useUndoRedo === true) {
     // somethign special
   } else {
     return next(call)
  }
}

// .. on a specific model action:
.actions(self => (
   function f() {}

   return {
      f: decorate(undoRedo, f)
   }
})

// .. on the 'root' model of your tree:
.actions(self => ({
   postCreate() {
       // intercept all undo/redo using actions in the store
       addMiddleware(self, undoRedoManager)
   }
})
```

Also see branch `action-improvements`, where this is WIP but should basically already work

**Comment by xaviergonz** [comment]
nice, wouldn't it be better though to use call.metadata.useUndoRedo = true so we don't end up modifying a variable from call that we were not supposed to change by mistake?

**Comment by xaviergonz** [comment]
ok, I saw that the new version supports decorators to add middlewares just for certain actions (awesome!).
My next question now however is, is there a way to pass a data object to that middleware? for example, something like:
```js
const requiredPermissions = {
  admin: true,
  // whatever else
};

const checkPermissionsMware = (actionCall, next, data) => {
  if (data.admin && !currentPermissions.admin) {
    // log something and ignore action, or throw an error
  }
  else {
    next(actionCall);
  }
};

decorate(checkPermissionsMware, someActionOnlyDoableByAdmins, requiredPermissions);
```

**Comment by xaviergonz** [comment]
Actually now that I think about it nevermind, I can just do this:

´´´js
const requiredPermissions = {
  admin: true,
  // whatever else
};

const createCheckPermissionsMware = (data: Permissions) => {
  return (actionCall, next) => {
    if (data.admin && !currentPermissions.admin) {
      // log something and ignore action, or throw an error
    }
    else {
      next(actionCall);
    }
  };
};

decorate(createCheckPermissionsMware(requiredPermissions), someActionOnlyDoableByAdmins);
´´´

**Comment by mattiamanzati** [comment]
Released with decorate on 0.11
